src/server/rendering/preact_hooks.ts

```diff
@@ -0,0 +1,287 @@
+import {
+  Component,
+  type ComponentChildren,
+  Fragment,
+  h,
+  type Options as PreactOptions,
+  options as preactOptions,
+  type VNode,
+} from "preact";
+import { assetHashingHook } from "../../runtime/utils.ts";
+import { renderToString } from "preact-render-to-string";
+import { RenderState } from "./state.ts";
+import { Island } from "../types.ts";
+
+// These hooks are long stable, but when we originally added them we
+// weren't sure if they should be public.
+interface AdvancedPreactOptions extends PreactOptions {
+  /** Attach a hook that is invoked after a tree was mounted or was updated. */
+  __c?(vnode: VNode, commitQueue: Component[]): void;
+  /** Attach a hook that is invoked before a vnode has rendered. */
+  __r?(vnode: VNode): void;
+  errorBoundaries?: boolean;
+  /** before diff hook */
+  __b?(vnode: VNode): void;
+}
+const options = preactOptions as AdvancedPreactOptions;
+
+// Enable error boundaries in Preact.
+options.errorBoundaries = true;
+
+// Set up a preact option hook to track when vnode with custom functions are
+// created.
+let current: RenderState | null = null;
+// Keep track of which component rendered which vnode. This allows us
+// to detect when an island is rendered within another instead of being
+// passed as children.
+let ownerStack: VNode[] = [];
```

Comment:
This appears to still be global - won't this cause any problems?

Reply:
We cannot completely remove all globals unless we add some sort of render context API to Preact. This one here in particular is always replaced when a new rendered is triggered here https://github.com/denoland/fresh/pull/1638/files#diff-85a0f65cd2b2187632bf4c5c3c0385eae00d0d0cd49feac175ca59558b02490fR50